From 7f67a4ea9d73f04ca9796b203a057bc517a15071 Mon Sep 17 00:00:00 2001
From: Zhou Guanghui <zhouguanghui1@huawei.com>
Date: Mon, 14 Nov 2022 13:28:28 +0000
Subject: [PATCH] mm: fix alloc CDM node memory for MPOL_BIND
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 7f67a4ea9d73f04ca9796b203a057bc517a15071
Modified-by-SEL: No


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I612UG
CVE: NA

------------------------------------------------------

Memory can be allocated from a specified CDM node only when it is
allowed to apply for memory from the CDM node. Otherwise, memory
will be allocated from other non-CDM nodes that are not allowed by
th cpuset.

Signed-off-by: Zhou Guanghui <zhouguanghui1@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 mm/page_alloc.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index 2feb99a0b98f..0965e38a4247 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -3867,7 +3867,8 @@ get_page_from_freelist(gfp_t gfp_mask, unsigned int order, int alloc_flags,
 			(alloc_flags & ALLOC_CPUSET) &&
 			!__cpuset_zone_allowed(zone, gfp_mask)
 #ifdef CONFIG_COHERENT_DEVICE
-			&& !(alloc_flags & ALLOC_CDM)
+			&& (!is_cdm_node(zone->zone_pgdat->node_id) ||
+			    !(alloc_flags & ALLOC_CDM))
 #endif
 		)
 				continue;
-- 
2.33.0

